File: python/callwire/codec.py
import msgpack


def pack_request(id: int, func: str, args: list, stream: bool = False) -> bytes:
    msg = {"id": id, "type": "request", "func": func, "args": args}
    if stream:
        msg["stream"] = True
    return msgpack.packb(msg)


def pack_response(id: int, result) -> bytes:
    return msgpack.packb({"id": id, "type": "response", "result": result})


def pack_error(id: int, error_type: str, message: str) -> bytes:
    return msgpack.packb(
        {"id": id, "type": "error", "error_type": error_type, "message": message}
    )


def pack_stream_chunk(id: int, result) -> bytes:
    return msgpack.packb({"id": id, "type": "stream_chunk", "result": result})


def pack_stream_end(id: int) -> bytes:
    return msgpack.packb({"id": id, "type": "stream_end"})


def pack_stream_chunk(id: int, result) -> bytes:
    return msgpack.packb({"id": id, "type": "stream_chunk", "result": result})


def pack_stream_close(id: int) -> bytes:
    return msgpack.packb({"id": id, "type": "stream_close"})


def pack_bidi_request(id: int, func: str, args: list = None) -> bytes:
    if args is None:
        args = []
    return msgpack.packb({"id": id, "type": "request", "func": func, "args": args, "stream": True})


def unpack(payload: bytes) -> dict:
    # raw=False decodes str properly — don't drop this flag
    #
    # Known edge: Python None unpacks into Go as the type's zero value (0 for int).
    # See SPEC.md type mapping table for details.
    return msgpack.unpackb(payload, raw=False)

File: go/callwire/codec.go
package callwire

import "github.com/vmihailenco/msgpack/v5"

type wireMessage struct {
	ID        uint64        `msgpack:"id"`
	Type      string        `msgpack:"type"`
	Func      string        `msgpack:"func,omitempty"`
	Args      []interface{} `msgpack:"args"`
	Stream    bool          `msgpack:"stream,omitempty"`
	Result    interface{}   `msgpack:"result,omitempty"`
	ErrorType string        `msgpack:"error_type,omitempty"`
	Message   string        `msgpack:"message,omitempty"`
}

func EncodeRequest(id uint64, funcName string, args []interface{}) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "request", Func: funcName, Args: args})
}

func encodeRequest(id uint64, funcName string, args []interface{}) ([]byte, error) {
	return EncodeRequest(id, funcName, args)
}

func encodeStreamRequest(id uint64, funcName string, args []interface{}) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "stream_request", Func: funcName, Args: args})
}

func encodeResponse(id uint64, result interface{}) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "response", Result: result})
}

func encodeError(id uint64, errorType, message string) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "error", ErrorType: errorType, Message: message})
}

func encodeStreamChunk(id uint64, result interface{}) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "stream_chunk", Result: result})
}

func encodeStreamEnd(id uint64) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "stream_end"})
}

func encodeStreamClose(id uint64) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "stream_close"})
}

func encodeBidiRequest(id uint64, funcName string, args []interface{}) ([]byte, error) {
	return msgpack.Marshal(wireMessage{ID: id, Type: "request", Func: funcName, Args: args, Stream: true})
}

func decodeMessage(payload []byte) (wireMessage, error) {
	var msg wireMessage
	err := msgpack.Unmarshal(payload, &msg)
	return msg, err
}
